include/dlaf/communication/async.h

```diff
@@ -0,0 +1,108 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2019, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+#pragma once
+
+/// @file
+
+#include <utility>
+#include <vector>
+
+#include <mpi.h>
+#include <hpx/async.hpp>
+#include <hpx/include/parallel_executors.hpp>
+#include <hpx/include/threads.hpp>
+#include <hpx/lcos/future.hpp>
+#include <hpx/util/yield_while.hpp>
+
+#ifdef DLAF_WITH_MPI_FUTURES
+#include <hpx/mpi/mpi_executor.hpp>
+#include <hpx/mpi/mpi_future.hpp>
+#endif
+
+namespace dlaf {
+namespace comm {
+
+/// The function only accepts non-blocking MPI routines. The MPI library needs
+/// to support MPI_THREAD_MULTIPLE.
+///
+/// Example usage:
+///
+/// dlaf::comm::async(ex, comm, MPI_Irecv, recv_ptr, num_recv_elems, MPI_DOUBLE, src_rank, tag);
+///
+/// compared to the usual MPI non-blocking call:
+///
+/// MPI_Irecv(recv_ptr, num_recv_elems, MPI_DOUBLE, src_rank, tag, comm, &request);
+///
+/// there are two differences
+///
+/// 1) MPI_Comm (comm) goes after the executor in async()
+/// 2) MPI_Requst (request) is omitted as it is handled internally in async()
+///
+/// Note that if starting the communication depends on a previous task, it is often a good idea to call
+/// the `get()` method on a returned future to avoid weird constructs like futute<future<>>.
+///
+/// Example:
+///
+/// auto comm_fut = prev_fut.then([=] {
+///   return dlaf::comm::async(...).get();
+/// });
+///
+/// or don't return the future from `dlaf::comm::async()` and just do:
+///
+/// auto comm_fut = prev_fut.then([=] {
+///   dlaf::comm::async(...);
+/// });
+///
+template <class F, class... Ts>
+hpx::future<void> async(hpx::threads::executors::pool_executor& ex, MPI_Comm comm, F&& f,
```

Comment:
I see you are still working on it and you have to make it compatible with the `Data`/`Message` API.

Just an additional suggestion. I would make this an `internal` function, and I would keep the same functions names used in `dlaf::comm:sync`, specifying two functions `dlaf::comm::async::send_to` and `dlaf::comm::async::receive_from`.

https://github.com/eth-cscs/DLA-Future/blob/ad7962e9073b883f8d527c34831fe0b85b27901a/include/dlaf/communication/sync/basic.h#L23-L42

Reply:
Hmm, this one I will have think about for a bit longer. Note that the `dlaf::comm::async()` can schedule any non-blocking routine, not only Isend and Irecv (e.g. Ireduce etc). The difficulty with integrating the Data/Message API is deciding where to place the parameters for the message size, type and buffer encoded in the API. The current API exploits the fact that MPI_Comm and MPI_Request are always at the end, the Data/Message API parameters are somewhere in the middle.

Reply:
We can, of course, just have separate function, like we do now : `send_to()` / `receive_from()`, etc. Perhaps that is what you meant. In that case, yeah, there won't be a problem.

Reply:
Yes, what you are saying in the last comment it is exactly what I meant.

About your problem with `Data`/`Message` API, the additional layer of indirection (e.g. `send_to`,`receive_to`,...) before the `internal` generic function may help you in managing it. If I get correctly your point, you can leave as it is the `internal` generic function, and then you will have different "proxy" functions with different sets of parameters (potentially with a different order then the one you need for the internal one).